guidedremediation/internal/manifest/python/poetry.go

```diff
@@ -0,0 +1,187 @@
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.PyPI,
+					Name:   d.Name,
+				},
+				Version:     d.Constraint,
+				VersionType: resolve.Requirement,
+			},
+			Type: dt,
+		})
+	}
+	return reqs
+}
+
+// Read parses the manifest from the given file.
+func (r poetryReadWriter) Read(path string, fsys scalibrfs.FS) (manifest.Manifest, error) {
+	path = filepath.ToSlash(path)
+	f, err := fsys.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var proj pyProject
+	if _, err := toml.NewDecoder(f).Decode(&proj); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal pyproject.toml: %w", err)
+	}
+
+	allReqs := []resolve.RequirementVersion{}
+	groups := make(map[manifest.RequirementKey][]string)
+
+	// Dependencies
+	allReqs = append(allReqs, parseDependencies(proj.Project.Dependencies, false)...)
+
+	// Optional dependencies
+	for groupName, deps := range proj.Project.OptionalDependencies {
+		groupReqs := parseDependencies(deps, true)
+		allReqs = append(allReqs, groupReqs...)
+		for _, r := range groupReqs {
+			key := manifest.RequirementKey(r.VersionKey.PackageKey)
+			groups[key] = append(groups[key], groupName)
+		}
+	}
+
+	return &pythonManifest{
+		filePath: path,
+		root: resolve.Version{
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.PyPI,
+					Name:   proj.Project.Name,
+				},
+				VersionType: resolve.Concrete,
+				Version:     proj.Project.Version,
+			},
+		},
+		requirements: allReqs,
+		groups:       groups,
+	}, nil
+}
+
+// Write writes the manifest after applying the patches to outputPath.
+func (r poetryReadWriter) Write(original manifest.Manifest, fsys scalibrfs.FS, patches []result.Patch, outputPath string) error {
+	return write(fsys, original.FilePath(), outputPath, patches, updatePyproject)
+}
+
+// updatePyproject takes an io.Reader representing the pyproject.toml file
+// and a map of package names to their new version constraints, returns the
+// file with the updated requirements as a string.
+func updatePyproject(reader io.Reader, requirements []TokenizedRequirements) (string, error) {
+	data, err := io.ReadAll(reader)
+	if err != nil {
+		return "", fmt.Errorf("error reading requirements: %w", err)
+	}
+	content := string(data)
+
+	var proj pyProject
+	if _, err := toml.Decode(content, &proj); err != nil {
+		return "", fmt.Errorf("failed to unmarshal pyproject.toml: %w", err)
+	}
+
+	updateDeps := func(deps []string) {
+		for _, req := range deps {
+			d, err := pypi.ParseDependency(req)
+			if err != nil {
+				log.Warnf("failed to parse Python dependency %s: %v", req, err)
+				continue
+			}
+
+			newReq, ok := findTokenizedRequirement(requirements, d.Name, tokenizeRequirement(d.Constraint))
+			if ok {
+				updatedReq := replaceRequirement(req, newReq)
+				content = strings.Replace(content, `"`+req+`"`, `"`+updatedReq+`"`, 1)
+			}
+		}
+	}
```

Comment:
This is a pretty clever workaround!
I just hope poetry requirements are always the same as lines in requirements.txt